File: day10/scala/src/main/scala/advent10.scala
package advent

import scala.io.Source

case class Point(x: Int, y: Int) {
  def `+`(that: Point) = copy(x = x + that.x, y = y + that.y)
}

case class Star(position: Point, velocity: Point) {
  def next = copy(position = position + velocity)
}

case class Stars(stars: List[Star]) {
  override def toString = {
    val (min, max) = gridDims
    (for(y <- (min.y to max.y)) yield {
      (for(x <- min.x to max.x) yield {
        if(stars.exists(s => s.position.x == x && s.position.y == y)) "#" else "."
      }).mkString
    }).mkString("\n")
  }

  lazy val gridDims: (Point, Point) = {
    val minx = stars.map(_.position.x).min
    val miny = stars.map(_.position.y).min
    val maxx = stars.map(_.position.x).max
    val maxy = stars.map(_.position.y).max
    (Point(minx, miny), Point(maxx, maxy))
  }

  lazy val size: Point = {
    val (min, max) = gridDims
    Point(max.x - min.x, max.y - min.y)
  }

  def next = copy(stars = stars.map(_.next))
}

object Stars {
  def fromString(s: String) =
    Stars(s.split("\n").map(Star.fromString _).toList)
}

object Point {
  val Pat = """\s*(-?[0-9]+)\s*,\s*(-?[0-9]+)""".r
  def fromString(s: String): Point = s match {
    case Pat(x, y) => Point(Integer.parseInt(x), Integer.parseInt(y))
  }
}

object Star {
  val Pat = """position=<(.+)>\s+velocity=<(.+)>""".r
  def fromString(s: String): Star = s match {
    case Pat(position, velocity) => Star(Point.fromString(position), Point.fromString(velocity))
  }
}

object Advent10 {

  @annotation.tailrec
  def solve(stars: Stars, last: Stars): Stars = {
    if(stars.size.x < last.size.x && stars.size.y < last.size.y)
      solve(stars.next, stars)
    else
      last
  }

  def main(args: Array[String]): Unit = {
    val inputFile = args.headOption.getOrElse("../example.txt")
    val stars = Stars.fromString(Source.fromFile(inputFile).mkString)
    println(solve(stars.next, stars))
  }
}

File: day10/advent10.rb
class Point
  attr_reader :x, :y
  def initialize(x, y)
    @x = x; @y = y
  end
  def +(other)
    Point.new(@x + other.x, @y + other.y)
  end
  def self.from_s(s)
    x, y = s.split(/\s*,\s*/)
    Point.new(x.to_i,y.to_i)
  end
end

class Star
  attr_reader :pos, :vel
  def initialize(pos, vel)
    @pos = pos; @vel = vel
  end
  def next
    Star.new(pos + vel, vel)
  end
  def self.from_s(s)
    m = /.*<(.*)>.*<(.*)>/.match(s)
    Star.new(Point.from_s(m[1]), Point.from_s(m[2]))
  end
end

class Stars
  include Enumerable
  attr_reader :size, :dims
  def initialize(stars)
    @stars = stars
    pos = @stars.map(&:pos)
    xs = pos.map(&:x)
    ys = pos.map(&:y)
    @dims = [ Point.new(xs.min, ys.min),
      Point.new(xs.max, ys.max) ]
    min, max = dims
    @size = Point.new(max.x - min.x, max.y - min.y)
  end
  def self.from_lines(lines)
    Stars.new(lines.map { |l| Star.from_s(l) }.to_a)
  end
  def next
    Stars.new(@stars.map(&:next))
  end
  def to_s
    min, max = dims
    (min.y..max.y).map do |y|
      (min.x..max.x).map do |x|
        if @stars.detect { |s| s.pos.x == x and s.pos.y == y } then "#" else " " end
      end.join
    end.join("\n")
  end
end

init_pos = Stars.from_lines(
  File.readlines(ARGV[0] || "example.txt")
)

stars = init_pos.next
last = init_pos
turn = 0
while stars.size.x < last.size.x and stars.size.y < last.size.y
  turn += 1
  last = stars
  stars = stars.next
end

puts last
puts turn
